pkg/providers/vsphere/vmprovider_vm.go

```diff
@@ -1053,6 +1067,97 @@ func (vs *vSphereVMProvider) getTags(
+	//    function in the file vmlifecycle/update_status.go).
+	//
+	// 2. If no such entry exists, then add one. Ensure the information
+	//    about the volume matches the disk's current information, such
+	//    as its controller key, controller type, unit number, etc.
+	//
+	//    If no entry existed, then use random value (that is DNS safe)
+	//    for the value of the claimName field.
+	//
+	// 3. Use the controllerutil.CreateOrPatch function to ensure a
+	//    PersistentVolumeClaim object exists for this disk. The PVC
+	//    *must* have a spec.dataSourceRef set that points to this VM.
+	//    See https://kubernetes.io/docs/concepts/storage/persistent-volumes/#using-volume-populators
+	//    for more information on this field.
+	//
+	//    The name of this PVC *must* match the value of
+	//    spec.volumes[].persistentVolumeClaim.claimName from the prior
+	//    step.
+	//
+	//    The PVC should also have an OwnerRef set that points back to
+	//    this VirtualMachine object. If the prior step created a *new*
+	//    entry in spec.volumes, then this OwnerRef should be a
+	//    ControllerOwnerRef.
+	//
+	//    If the PVC does not have a ControllerOwnerRef that points to
+	//    this VM, then the PVC's storageClass,
+	//    resources.requests.storage, and resources.limits.storage
+	//    fields should have their values set to match the analogous
+	//    values of the current disk if those fields's values are empty.
+	//    The same goes for its crypto-related information (TODO -- find
+	//    the PVC annotation that specifies the BYOK EncryptionClass).
+	//
+	//    The use of the ControllerOwnerRef allows us to determine if the
+	//    PVC was created by VM Operator or by an external actor, such as
+	//    a DevOps user or Blueprint service. In the case of the latter,
+	//    they may pre-create the PVC object with the dataSourceRef field
+	//    pointing to this VM in order to ensure the unmanaged volume is
+	//    migrated to a new storage class and/or resized when being
+	//    converted to a managed volume (PVC). Therefore, we only want to
+	//    set those fields if VM Op was the one to create the PVC in the
+	//    first place.
+	//
+	// 4. If the PVC object's status.phase field is empty, then use the
+	//    controllerutil.CreateOrPatch function to ensure a
+	//    CnsRegisterVolume object exists for this disk.
+	//
+	//    See https://docs.google.com/document/d/1L6hGuRMY2Caci5OZyZMxp84Bpquw-WCX2ZQQ_2d5mlE/edit?usp=sharing
+	//    for more details on how CnsRegisterVolume is being augmented.
+	//
+	//    The CnsRegisterVolume field pvcName should match the name of
+	//    the PVC created/patched in the previous step.
+	//
+	//    The CnsRegisterVolume object should have an OwnerRef set to
+	//    this VM.
+	//
+	//    The CnsRegisterVolume object should have a label, ex.
+	//    vmoperator.vmware.com/created-by, with a value that is this
+	//    VM's name.
+	//
+	// 5. If the CnsRegisterVolume status indicates the operation is
```

Comment:
@xing-yang , can we please have a `TtlAfterFinished` style garbage collection for CnsRegisterVolume resources?  Once this API is successful, it is not useful for consumers.  So, it will be a good value add.  RegisterVM API also has to handle the cleanup, so it will benefit there as well.